CAMEL-10447 Add contract based type awareness and transformer which l…

igarashitm opened the pull request:
…everages the type metadata

https://issues.apache.org/jira/browse/CAMEL-10447

davsclaus commented:
I think this PR is too invasive and touches to many places.

It should really only be about route IN and OUT, and not for EIPs like Send / ProducerCache and many other places.

Also it seems contract was added to the ProcessorDefinition which is the base EIP building block which then adds it into the DSL for all EIPs which is not what we want.

Also the DSL names should not contain dot in their name, eg transformer.dataFormat etc.

And what is the reason for the urn syntax using { } which confuses users. Its maybe better to come up with a better style, as there can be many type mistakes with those horrible XML Namespace like syntaxes.

Also the `<from>` should not contain child nodes, that is not the style we do in Camel, it would be

   from
      inputType
      outputType
      to a
      to b

igarashitm commented:
Hi @davsclaus , thank you for the review!

 Do you think it should only be available on endpoint aware processor like to, wireTap and etc, or it shouldn't be in producer side at all? The idea on having contract in ProcessorDefinition is that, to declare the contract also on producer side:
```xml
<route id="a">
    <from uri="direct:a">
        <inputType urn="a-in">
        <outputType urn="a-out">
    </from>
    <multicast>
        <inputType urn="x-in">
        <outputType urn="x-out">
        <to uri="direct:b"/>
        <to uri="file:aout"/>
    </multicast>
<route id="b">
    <from uri="direct:b">
        <inputType urn="b-in">
        <outputType urn="b-out">
    </from>
</route>
```
This allows to declare a contract on both of producer and consumer side, and if it's different then transformer would be applied. In above example input 'x-in' is transformed into 'b-in' when routing from 'direct:a' to 'direct:b'. But in this way I needed to put those element as a child node to identify which is the element the contract belongs to when persing XML DSL. For Java DSL the 'scopedContractAware' keeps the ProcessorDefinition which is added right before.
https://github.com/apache/camel/pull/1252/files#diff-a23471776349b96a76ce759ce2bf3db9R115

For the transformer.dataFormat kind of naming, OK I'll change it like dataFormats is doing. I should have done this when I introduced 'transformers' parent element.

For the XML namespace like syntax, I'll remove that namespace style example from test code and example. Right now the type URN could be arbitrary string literal, so it doesn't have any restriction. Some convention may be required for that at some point, but I think it must follow what component metadata look like in Camel 3.x.

igarashitm commented:
I almost forgot I couldn't use 'dataFormat', 'endpoint' and 'bean' as it's already reserved. Renamed like from 'transformer.dataFormat' to 'dataFormatTransformer' instead.

davsclaus commented:
We agreed on that it should only be for routes where you can define what input and output a route receives and returns.

Look at how the rest-dsl setup its binding which is similar to this. 

The Camel users want something simple where they can configure on their routes what the get as IN and output as OUT. Also in a loosly fashion to just define the type as xml, json, java etc. For java you may point to Java classes, but just saying java itself helps Camel developers to understand this a bit more.

Also a good idea is to build an example in pure Java such as springboot or cdi, that helps you build the DSL in a way it works well there too. The example is only XML and it hides sometimes how you may need to do some things a bit different in java, eg configure a data format etc.

igarashitm commented:
I'm still trying to fix the code and need to add more example though, I got a question...

So I tried to use Exchange.addOnCompletion() to process OUT message, but it turned out it doesn't work well for contract processing if there's nested internal route. For example following route needs to add 2 onCompletion(), but in opposite order, which is not possible as onCompletion() callback is invoked in registered order.
```java
from("direct:a")
    .inputType("java:ain")
    .outputType("java:aout")
    .to("direct:b");
from("direct:b")
    .inputType("java:bin")
    .outputType("java:bout")
    ....
```
Instead I added contract processing class as a CamelInternalProcessorAdvice.
https://github.com/apache/camel/pull/1252/files#diff-c4c38676db3f1690554e0dcf4d81233eR198

I wonder if it looks OK for you @davsclaus ?

Thanks,
Tomo

davsclaus commented:
We should then avoid the double wrapping, eg try to add the advice on the existing internal processor.

Also mind that input and output is optional, so you can define a route with on an input contract. Or with only an output contract.

igarashitm commented:
Thank you for the comment! Yes I changed to just add the advice on existing CamelInternalProcessor. And both of inputType and outputType are optional, so that we can specify only one of these. I'll add those testcases as well.

igarashitm commented:
I got those fixed - if you could have a chance to look and make more feedback, that would be really appreciated!
* Moved inputType/outputType to RouteDefinition from ProcessorDefinition. Now it only works for route level. Also it's no longer a child of from or other processor in XML. It's right under route.
  * https://github.com/apache/camel/pull/1252/files#diff-796677347044b03ab8cd692feda41aaa
* DefaultRouteContext.commit() adds ContractAdvice on CamelInternalProcessor directly which performs contract based transformation instead of wrapping with ContractProcessor
  * https://github.com/apache/camel/pull/1252/files#diff-c4c38676db3f1690554e0dcf4d81233e
  * https://github.com/apache/camel/pull/1252/files#diff-b093deeaf97c04e9def03f1d4a6f7723
* Added a few tests verifying inputType only or outputType only works. Also verifying short type name usage which only has its data model like "xml", "json" or "java"
  * https://github.com/apache/camel/pull/1252/files#diff-5be863f06b87b852fdf81d09ee139274
* Added 2 more simpler examples which only trigger a xslt endpoint transformation. camel-example-transformer-cdi and camel-example-transformer-blueprint are doing same routing, but with different DSL. The first example camel-example-transformer-demo which demonstrates a bit more transformation in Spring XML is still available
  * [Blueprint] - https://github.com/apache/camel/pull/1252/files#diff-e0cbeccf0006c906cfd0535bf0a91e75 
  * [CDI] - https://github.com/apache/camel/pull/1252/files#diff-6898716fca98b99755032e172d3debcb
  * [Demo(Spring)] - https://github.com/apache/camel/pull/1252/files#diff-1b27a607c8c318643b270f6916cdec19

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/RouteDefinition.java:
```diff
@@ -629,6 +632,50 @@ public RouteDefinition shutdownRunningTask(ShutdownRunningTask shutdownRunningTa
+     * @param name input type URN
+     * @return the builder
+     */
+    public RouteDefinition inputType(String name) {
```
name parameter should be urn

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/RouteDefinition.java:
```diff
@@ -629,6 +632,50 @@ public RouteDefinition shutdownRunningTask(ShutdownRunningTask shutdownRunningTa
+     * @param name output type URN
+     * @return the builder
+     */
+    public RouteDefinition outputType(String name) {
```
name parameter should be urn

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/OutputTypeDefinition.java:
```diff
@@ -0,0 +1,80 @@
+     * Set output type via Java Class.
+     * @param clazz Java Class
+     */
+    public void setJavaClass(Class<?> clazz) {
```
Maybe just set this as this.urn = "java:" + clazz.getName() then you dont need the transient java type

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/transformer/CustomTransformerDefinition.java:
```diff
@@ -0,0 +1,106 @@
+import org.apache.camel.spi.Transformer;
+import org.apache.camel.util.ObjectHelper;
+
+/**
```
The javadoc says bean transformer but the class is named custom

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/transformer/CustomTransformerDefinition.java:
```diff
@@ -0,0 +1,106 @@
+            if (transformerClass == null) {
+                throw new IllegalArgumentException("Cannot find transformer class: " + type);
+            }
+            transformer = transformerClass.newInstance();
```
Use camelContext injector to create a new instance as it can integrates with spring/blueprint etc and supports IoC stuff

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/transformer/DataFormatTransformerDefinition.java:
```diff
@@ -0,0 +1,161 @@
+import org.apache.camel.spi.RouteContext;
+import org.apache.camel.spi.Transformer;
+
+/**
```
AFAIR the javadoc is use for documenting what this DSL does. Can you put a bit better documentation for those.

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/transformer/TransformerDefinition.java:
```diff
@@ -0,0 +1,133 @@
+    private String from;
+    @XmlAttribute
+    private String to;
+    @XmlTransient
```
Avoid dependency in definition classes to a specify CamelContext instance

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/transformer/TransformerDefinition.java:
```diff
@@ -0,0 +1,133 @@
+        return doCreateTransformer();
+    };
+
+    protected abstract Transformer doCreateTransformer() throws Exception;
```
Add (CamelContext context) as parameter instead and remove the local CamelContext instance

davsclaus commented on camel-core/src/main/java/org/apache/camel/processor/CamelInternalProcessor.java:
```diff
@@ -854,4 +858,120 @@ public void after(Exchange exchange, Object data) throws Exception {
+            }
+        }
+        
+        private static void convertBody(Message message, DataType from, DataType to) throws Exception {
```
Can we split this method into smaller methods where each method does one kind of conversion

davsclaus commented on camel-core/src/main/java/org/apache/camel/spi/RouteContext.java:
```diff
@@ -192,4 +192,9 @@
+     * 
+     * @param contract
+     */
+    void setContract(Contract contract);
```
javadoc missing

davsclaus commented on camel-core/src/test/java/org/apache/camel/impl/transformer/TransformerContractTest.java:
```diff
@@ -0,0 +1,173 @@
+import org.apache.camel.spi.RouteContext;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
will cause checkstyle error

davsclaus commented on camel-core/src/main/java/org/apache/camel/impl/DefaultCamelContext.java:
```diff
@@ -276,6 +280,8 @@
     private Date startDate;
     private ModelJAXBContextFactory modelJAXBContextFactory;
+    private List<TransformerDefinition> transformers = new ArrayList<TransformerDefinition>();
+    private Map<TransformerKey, Transformer> transformerRegistry = new HashMap<TransformerKey, Transformer>();
```
I think we should create a TransformerRegistry class like we have EndpointRegistry. Then we can also use that to expose JMX information at runtime etc. 

https://github.com/apache/camel/blob/master/camel-core/src/main/java/org/apache/camel/spi/EndpointRegistry.java

https://github.com/apache/camel/blob/master/camel-core/src/main/java/org/apache/camel/management/mbean/ManagedEndpointRegistry.java

davsclaus commented on examples/camel-example-transformer-cdi/src/main/java/org/apache/camel/example/transformer/cdi/MyRoutes.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void configure() {
+        EndpointTransformerDefinition eptd = new EndpointTransformerDefinition();
```
We should ponder to make this Java DSL nicer to use, having to create an Definition class and configure it is a bit cumbersome.

It could be something alike

    transform().from("xxx").to("yyy").withUrl("zzz").withJava(MyClass).withDataFormat(myDataFormat) 

or something, where you can chose any of those 3 kinds url, java or data format. Yeah it takes time to come up with a good dsl.

davsclaus commented on camel-core/src/main/java/org/apache/camel/impl/transformer/ProcessorTransformer.java:
```diff
@@ -0,0 +1,122 @@
+        }
+        
+        LOG.debug("Sending to transform processor '{}'", processor);
+        DefaultExchange transformExchange = new DefaultExchange(exchange);
```
Not sure if we need to create a new exchange, we can send the existing exchange.

davsclaus commented on camel-core/src/main/java/org/apache/camel/impl/transformer/ProcessorTransformer.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Override
+    protected void doStart() throws Exception {
+        if (this.processor instanceof ServiceSupport) {
```
Use ServiceHelper.startService / and ditto for stop. No need for instanceof check either

davsclaus commented:
@igarashitm yeah lets talk about that new exchange thingy later. There is plenty of code changes already which looks good. So I will attempt to get this merged, then you/we can refine this over the following weeks.

Do you mind writing a documentation file (adoc) about this new functionality at
https://github.com/apache/camel/tree/master/camel-core/src/main/docs

You can do that in a new PR but wait for this PR to get merged

igarashitm commented:
Thank you that sounds like a plan. I still have following 3 things need to be done, I'll also create another PR for them.
* Split doTransform() method into smaller methods
* Create TransformerRegistry and JMX bits
* Fluent builder for Transformer definition

davsclaus commented:
Thanks for the PR it has been merged. Do you mind closing this?

igarashitm commented:
Thank you @davsclaus so much for your big support to get this improved and finally merged! I'll continue on this to make it better. closing
